pkg/ipcache/ipcache.go

```diff
@@ -806,18 +807,46 @@ func (ipc *IPCache) LookupByPrefix(IP string) (Identity, bool) {
 	return ipc.LookupByPrefixRLocked(IP)
 }
 
+// LookupSecIDByIPLocal performs a longest prefix match lookup in the IPCache for
+// the identity corresponding to the specified address (or, in the case of no
+// direct match, any shorter prefix). Returns the corresponding identity and
+// whether the IP is local and whether a match was found.
+func (ipc *IPCache) LookupSecIDByIPLocal(ip netip.Addr) (id Identity, local bool, ok bool) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+
+	id, ok = ipc.lookupSecIDByIPRLocked(ip)
+	if ok {
+		hostIP, _ := ipc.getHostIPCache(ip.String())
+		if ip4 := hostIP.To4(); ip4 != nil {
+			local = ip4.Equal(node.GetIPv4())
+		} else if hostIP != nil {
+			local = hostIP.Equal(node.GetIPv6())
+		}
```

Comment:
So the "hostIP" here is the tunnel IP, right? We don't always set the tunnel IP, depending on IP type, encryption, hostfirewall etc:

https://github.com/cilium/cilium/blob/076653691d0d15b4449d6e2791bdbaea5298ed5f/pkg/node/manager/manager.go#L389-L396

I don't think relying on the tunnel IP is robust. We should instead rely on the IP identity (i.e. is it localhost) and on the endpoint list.

Reply:
You are right, and as you suggested elsewhere, checking against the reserved host identity suffices here and is much simpler!